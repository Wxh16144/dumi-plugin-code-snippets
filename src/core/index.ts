export { default as remarkPlugin } from './remarkPlugin';

export type { IProps as RemarkPluginProps } from './remarkPlugin';
