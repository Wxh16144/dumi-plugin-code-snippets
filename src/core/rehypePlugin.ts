import { unistUtilVisit } from 'dumi';

function rehypePlugin() {
  return (tree: any) => {
    unistUtilVisit.visit(tree, 'element', (node, index, parent) => {

    });
  };
}

export default rehypePlugin;
