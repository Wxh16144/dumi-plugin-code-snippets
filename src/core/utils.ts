// Duplicate of: https://github.com/vuejs/vitepress/blob/f186901a5157c904b3593089d72f2bad3530e7a3/src/node/markdown/plugins/snippet.ts

/**
 * raw path format: "/path/to/file.extension#region {meta} [title]"
 *    where #region, {meta} and [title] are optional
 *    meta can be like '1,2,4-6 lang', 'lang' or '1,2,4-6'
 *    lang can contain special characters like C++, C#, F#, etc.
 *    path can be relative to the current file or absolute
 *    file extension is optional
 *    path can contain spaces and dots
 *
 * captures: ['/path/to/file.extension', 'extension', '#region', '{meta}', '[title]']
 */
export const rawPathRegexp =
  /^(.+?(?:(?:\.([a-z0-9]+))?))(?:(#[\w-]+))?(?: ?(?:{(\d+(?:[,-]\d+)*)? ?(\S+)?}))? ?(?:\[(.+)\])?$/;

export function rawPathToToken(rawPath: string) {
  const [filepath = '', extension = '', region = '', lines = '', lang = '', rawTitle = ''] = (
    rawPathRegexp.exec(rawPath) || []
  ).slice(1);

  const title = rawTitle || filepath.split('/').pop() || '';

  return { filepath, extension, region, lines, lang, title };
}

export function dedent(text: string): string {
  const lines = text.split('\n');

  const minIndentLength = lines.reduce((acc, line) => {
    for (let i = 0; i < line.length; i++) {
      if (line[i] !== ' ' && line[i] !== '\t') return Math.min(i, acc);
    }
    return acc;
  }, Infinity);

  if (minIndentLength < Infinity) {
    return lines.map((x) => x.slice(minIndentLength)).join('\n');
  }

  return text;
}

function testLine(line: string, regexp: RegExp, regionName: string, end: boolean = false) {
  const [full, tag, name] = regexp.exec(line.trim()) || [];

  return (
    full && tag && name === regionName && tag.match(end ? /^[Ee]nd ?[rR]egion$/ : /^[rR]egion$/)
  );
}

export function findRegion(lines: Array<string>, regionName: string) {
  const regionRegexps = [
    /^\/\/ ?#?((?:end)?region) ([\w*-]+)$/, // javascript, typescript, java
    /^\/\* ?#((?:end)?region) ([\w*-]+) ?\*\/$/, // css, less, scss
    /^#pragma ((?:end)?region) ([\w*-]+)$/, // C, C++
    /^<!-- #?((?:end)?region) ([\w*-]+) -->$/, // HTML, markdown
    /^#((?:End )Region) ([\w*-]+)$/, // Visual Basic
    /^::#((?:end)region) ([\w*-]+)$/, // Bat
    /^# ?((?:end)?region) ([\w*-]+)$/, // C#, PHP, Powershell, Python, perl & misc
  ];

  let regexp = null;
  let start = -1;

  // @ts-ignore
  for (const [lineId, line] of lines.entries()) {
    if (regexp === null) {
      for (const reg of regionRegexps) {
        if (testLine(line, reg, regionName)) {
          start = lineId + 1;
          regexp = reg;
          break;
        }
      }
    } else if (testLine(line, regexp, regionName, true)) {
      return { start, end: lineId, regexp };
    }
  }

  return null;
}

export function resolvePath(aliasMap: Record<string, string>, inputPath: string) {
  for (const alias in aliasMap) {
    if (inputPath.startsWith(alias)) {
      return inputPath.replace(alias, aliasMap[alias]);
    }
  }
  return inputPath;
}
