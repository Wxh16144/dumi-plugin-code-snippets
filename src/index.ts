import type { IApi } from 'dumi';
import type { RemarkPluginProps } from './core';
import { remarkPlugin } from './core';

export default (api: IApi) => {
  api.register({
    key: 'modifyConfig',
    stage: Infinity,
    fn: (memo: IApi['config']) => {
      const cloneExtraRemarkPlugins = memo.extraRemarkPlugins;

      memo.extraRemarkPlugins = [
        [
          remarkPlugin,
          {
            codeBlockMode: memo.resolve?.codeBlockMode,
            cwd: api.cwd,
          } as Required<RemarkPluginProps>,
        ],
        ...(Array.isArray(cloneExtraRemarkPlugins)
          ? cloneExtraRemarkPlugins
          : ([cloneExtraRemarkPlugins].filter(Boolean) as any)),
      ];

      return memo;
    },
  });
};
