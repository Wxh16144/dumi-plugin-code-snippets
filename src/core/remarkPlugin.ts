import { defineConfig, unistUtilVisit } from 'dumi';
import fs from 'fs-extra';
import path from 'path';
import type { Node, Parent } from 'unist';
import { dedent, findRegion, rawPathToToken } from './utils';

const CH = '<'.charCodeAt(0);

type IDumiUserConfig = ReturnType<typeof defineConfig>;

export interface IProps {
  codeBlockMode?: Required<IDumiUserConfig>['resolve']['codeBlockMode'];
}

function remarkPlugin(opt: IProps) {
  const { codeBlockMode = 'active' } = opt;

  return (tree: any, vFile: any) => {
    const codeSnippets: [Node, number, Parent | undefined][] = [];

    unistUtilVisit.visit(tree, 'paragraph', (node, index, parent) => {
      if (!Array.isArray(node.children) && node.children.length > 1) {
        return unistUtilVisit.SKIP;
      }

      const [child] = node.children;

      if (child.type !== 'text') {
        return unistUtilVisit.SKIP;
      }

      const { value } = child;
      const isValidPrefix = value
        .slice(0, 3)
        .split('')
        .every((char: string) => {
          return char.charCodeAt(0) === CH;
        });

      if (!isValidPrefix) {
        return unistUtilVisit.SKIP;
      }

      codeSnippets.push([node, index!, parent]);
    });

    for (const [node, index, parent] of codeSnippets) {
      const cloneNode: any = { ...node };
      const rawPath = cloneNode.children[0].value.slice(3).trim();
      // todo: alias
      // .replace(/^@/, srcDir)
      const { filepath, extension, region, lines } = rawPathToToken(rawPath);
      const regionName = region.slice(1);

      // fixme: resourcePath is undefined，需要 dumi 那边给出解决方案
      const currentFileAbsPath = path.resolve(vFile.data.frontmatter!.resourcePath ?? '');
      const src = path.isAbsolute(filepath)
        ? filepath
        : path.join(path.dirname(currentFileAbsPath), filepath);

      let content: string = '';

      const srcStats = fs.existsSync(src) && fs.lstatSync(src);
      const isAFile = srcStats && (srcStats.isFile() || srcStats.isSymbolicLink());

      if (!isAFile) {
        content = isAFile ? `Invalid code snippet option` : `Code snippet path not found: ${src}`;
      } else {
        content = fs.readFileSync(src, 'utf8');
      }

      if (regionName) {
        const lines = content.split(/\r?\n/);
        const region = findRegion(lines, regionName);

        if (region) {
          content = dedent(
            lines
              .slice(region.start, region.end)
              .filter((line: string) => !region.regexp.test(line.trim()))
              .join('\n'),
          );
        }
      }

      // ref: https://github.com/umijs/dumi/blob/e0a864462c4a3f778c5a677150a02ab451e4b99f/src/loaders/markdown/transformer/rehypeDemo.ts#L53-L59
      let codeBlockModeStr: string = '';
      switch (codeBlockMode) {
        case 'active':
          codeBlockModeStr += '| pure';
          break;
        case 'passive':
        default:
          codeBlockModeStr = '';
          break;
      }

      // ref: https://github.com/umijs/dumi/blob/e0a864462c4a3f778c5a677150a02ab451e4b99f/src/loaders/markdown/transformer/rehypeHighlightLine.ts#L31
      let linesStr: string = '';
      if (lines) {
        linesStr = `{${lines}}`;
      }

      // replace node
      parent?.children.splice(index!, 1, {
        type: 'code',
        lang: extension,
        meta: `${linesStr}${codeBlockModeStr}`,
        value: content,
      } as any);
    }
  };
}

export default remarkPlugin;
