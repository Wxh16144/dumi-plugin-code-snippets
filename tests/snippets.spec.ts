import { remark } from 'remark';
import { VFile } from 'vfile';

import { remarkPlugin as core } from '../src/core';

describe('snippet', () => {
  const generateFile = (content: string) =>
    new VFile({
      value: content.trim(),
      data: {
        frontmatter: {
          filename: './tests/__fixtures__/index.md',
        },
      },
      basename: 'index.md',
      extname: '.md',
    }) as unknown as string;

  it('基础功能', () => {
    expect(remark().use(core, {}).processSync(generateFile(`<<< ./say-hi.ts`)).toString())
      .toMatchInlineSnapshot(`
      "\`\`\`ts | pure
      console.log('Hello dumi-plugin-code-snippets!');

      export function doSomething() {
        return 1;
      }

      \`\`\`
      "
    `);
  });

  it('带高亮行', () => {
    expect(remark().use(core, {}).processSync(generateFile(`<<< ./say-hi.ts{1}`)).toString())
      .toMatchInlineSnapshot(`
      "\`\`\`ts {1}| pure
      console.log('Hello dumi-plugin-code-snippets!');

      export function doSomething() {
        return 1;
      }

      \`\`\`
      "
    `);
  });

  it('支持tsx', () => {
    expect(remark().use(core, {}).processSync(generateFile(`<<< ./My.tsx`)).toString())
      .toMatchInlineSnapshot(`
      "\`\`\`tsx | pure
      import * as React from 'react';

      export interface MyProps {
        slogan?: React.ReactNode;
      }

      // #region snippet
      function My(props: React.PropsWithChildren<MyProps>) {
        const { children } = props;
        return (
          <>
            <div className=\\"my-slogan\\">
              <p>魔法师正在进行最后的仪式，为您带来一项惊艳功能</p>
              <strong>TBD: The Brilliant Discovery!</strong>
            </div>
            {children}
            {/* This is my.tsx} */}
          </>
        );
      }
      // #endregion snippet

      export default My;

      \`\`\`
      "
    `);
  });

  it('支持 vscode region', () => {
    expect(remark().use(core, {}).processSync(generateFile(`<<< ./My.tsx#snippet`)).toString())
      .toMatchInlineSnapshot(`
      "\`\`\`tsx | pure
      function My(props: React.PropsWithChildren<MyProps>) {
        const { children } = props;
        return (
          <>
            <div className=\\"my-slogan\\">
              <p>魔法师正在进行最后的仪式，为您带来一项惊艳功能</p>
              <strong>TBD: The Brilliant Discovery!</strong>
            </div>
            {children}
            {/* This is my.tsx} */}
          </>
        );
      }
      \`\`\`
      "
    `);
  });

  it('支持 region + 高亮行', () => {
    expect(remark().use(core, {}).processSync(generateFile(`<<< ./My.tsx#snippet{4}`)).toString())
      .toMatchInlineSnapshot(`
      "\`\`\`tsx {4}| pure
      function My(props: React.PropsWithChildren<MyProps>) {
        const { children } = props;
        return (
          <>
            <div className=\\"my-slogan\\">
              <p>魔法师正在进行最后的仪式，为您带来一项惊艳功能</p>
              <strong>TBD: The Brilliant Discovery!</strong>
            </div>
            {children}
            {/* This is my.tsx} */}
          </>
        );
      }
      \`\`\`
      "
    `);
  });

  it('支持 css 文件', () => {
    expect(remark().use(core, {}).processSync(generateFile(`<<< ./style.css`)).toString())
      .toMatchInlineSnapshot(`
      "\`\`\`css | pure
      #App {
        color: red;
      }

      \`\`\`
      "
    `);
  });

  it.skip('路径错误返回错误信息', () => {
    expect(remark().use(core, {}).processSync(generateFile(`<<< ./aa/bb/style.css`)).toString())
      .toMatchInlineSnapshot(`
      "\`\`\`css | pure
      Code snippet path not found: /Users/wuxh/Code/SelfProject/dumi-plugin-code-snippets/tests/__fixtures__/aa/bb/style.css
      \`\`\`
      "
    `);
  });
});
