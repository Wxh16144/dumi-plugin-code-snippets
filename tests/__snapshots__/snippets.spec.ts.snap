// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`snippet > 带高亮行 1`] = `
"\`\`\`ts {1}| pure
console.log('Hello dumi-plugin-code-snippets!');

export function doSomething() {
  return 1;
}

\`\`\`
"
`;

exports[`snippet > 支持 css 文件 1`] = `
"\`\`\`css | pure
#App {
  color: red;
}

\`\`\`
"
`;

exports[`snippet > 支持 region + 高亮行 1`] = `
"\`\`\`tsx {4}| pure
function My(props: React.PropsWithChildren<MyProps>) {
  const { children } = props;
  return (
    <>
      <div className=\\"my-slogan\\">
        <p>魔法师正在进行最后的仪式，为您带来一项惊艳功能</p>
        <strong>TBD: The Brilliant Discovery!</strong>
      </div>
      {children}
      {/* This is my.tsx} */}
    </>
  );
}
\`\`\`
"
`;

exports[`snippet > 支持 vscode region 1`] = `
"\`\`\`tsx | pure
function My(props: React.PropsWithChildren<MyProps>) {
  const { children } = props;
  return (
    <>
      <div className=\\"my-slogan\\">
        <p>魔法师正在进行最后的仪式，为您带来一项惊艳功能</p>
        <strong>TBD: The Brilliant Discovery!</strong>
      </div>
      {children}
      {/* This is my.tsx} */}
    </>
  );
}
\`\`\`
"
`;

exports[`snippet > 支持tsx 1`] = `
"\`\`\`tsx | pure
import * as React from 'react';

export interface MyProps {
  slogan?: React.ReactNode;
}

// #region snippet
function My(props: React.PropsWithChildren<MyProps>) {
  const { children } = props;
  return (
    <>
      <div className=\\"my-slogan\\">
        <p>魔法师正在进行最后的仪式，为您带来一项惊艳功能</p>
        <strong>TBD: The Brilliant Discovery!</strong>
      </div>
      {children}
      {/* This is my.tsx} */}
    </>
  );
}
// #endregion snippet

export default My;

\`\`\`
"
`;
